### PR TITLE
[Merged by Bors] - feat(MorphismProperty/Presheaf): Define relatively representable morphisms

### PR DESCRIPTION
**Pull request by callesonne**
This PR defines relatively representable morphisms, and adds some basic API.

Co-authored-by: Joël Riou <rioujoel@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+lemma isPullback : IsPullback (hf.fst g) (yoneda.map (hf.snd g)) f g :=
+  (hf g).choose_spec.choose_spec.choose_spec
+
+@[reassoc]
+lemma w : hf.fst g ≫ f = yoneda.map (hf.snd g) ≫ g := (hf.isPullback g).w
+
+/-- Variant of the pullback square when the first projection lies in the image of yoneda. -/
+lemma isPullback' : IsPullback (yoneda.map (hf'.fst' g)) (yoneda.map (hf'.snd g)) f' g :=
+  (hf'.yoneda_map_fst' _) ▸ (hf' g).choose_spec.choose_spec.choose_spec
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:103]:
```suggestion
  (hf'.yoneda_map_fst' _) ▸ hf'.isPullback g
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+@[reassoc (attr := simp)]
+lemma lift_snd : hf.lift i h hi ≫ hf.snd g = h :=
+  yoneda.map_injective <| by simpa [lift] using PullbackCone.IsLimit.lift_snd _ _ _ _
+
+end
+
+section
+
+variable {Z : C} (i : Z ⟶ Y) (h : Z ⟶ X) (hi : (yoneda.map i) ≫ f' = yoneda.map h ≫ g)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:159]:
```suggestion
variable {Z : C} (i : Z ⟶ Y) (h : Z ⟶ X) (hi : yoneda.map i ≫ f' = yoneda.map h ≫ g)
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+/-!
+
+# Representable morphisms of presheaves
+
+In this file we define and develop basic results on the representability of morphisms of presheaves.
+
+## Main definitions
+
+* `Presheaf.representable` is a `MorphismProperty` expressing the fact that a morphism of presheaves
+  is representable.
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:18]:
```suggestion
* `Presheaf.representable` is a `MorphismProperty` expressing the fact that a morphism `f` of presheaves
  is representable, i.e. that any morphism from a representable presheaf to the target can be pulled back
  along `f` to a morphism from a representable presheaf to the source.
```
Feel free to rephrase, I just think there should be a short description here of what this means mathematically.

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+## API
+
+Given `hf : Presheaf.representable f`, with `f : F ⟶ G` and `g : yoneda.obj X ⟶ G`, we provide:
+* `hf.pullback g` which is the object in `C` such that `yoneda.obj (hf.pullback g)` forms a
+  pullback square of `f` and `g`.
+* `hf.snd g` is the morphism `hf.pullback g ⟶ X`
+* `hf.fst g` is the morphism `yoneda.obj (hf.pullback g) ⟶ F`
+*  Whenever `f` is of type `yoneda.obj Y ⟶ G`, we also have `hf.fst' g` which is the preimage
+   under `yoneda` of `hf.fst g`.
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:28]:
```suggestion
*  Whenever `f` is of type `yoneda.obj Y ⟶ G`, we also have `hf.fst' g : hf.pullback g ⟶ Y`
which is the preimage under `yoneda` of `hf.fst g`.
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+* `hom_ext`, `hom_ext'`, `lift`, `lift'` are variants of the universal property of
+  `yoneda.obj (hf.pullback g)`, where as much as possible have been formulated internally to `C`.
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:30]:
```suggestion
* `hom_ext`, `hom_ext'`, `lift`, `lift'` are variants of the universal property of
  `yoneda.obj (hf.pullback g)`, where as much as possible has been formulated internally to `C`.
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+* `symmetry` and `symmetryIso` are variants of the fact that pullbacks are symmetric for
+  representable morphisms, formulated internally to `C`.
+
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:34]: Is there a reason for this blank line between bullet points? If not, could you remove it?

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+noncomputable def lift : Z ⟶ hf.pullback g :=
+  Yoneda.fullyFaithful.preimage <| PullbackCone.IsLimit.lift (hf.isPullback g).isLimit _ _ hi
+
+@[reassoc (attr := simp)]
+lemma lift_fst : yoneda.map (hf.lift i h hi) ≫ hf.fst g = i := by
+  simpa [lift] using PullbackCone.IsLimit.lift_fst _ _ _ _
+
+
+@[reassoc (attr := simp)]
+lemma lift_snd : hf.lift i h hi ≫ hf.snd g = h :=
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:152]:
```suggestion

@[reassoc (attr := simp)]
lemma lift_snd : hf.lift i h hi ≫ hf.snd g = h :=
```

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean (hunk 3 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:18]: Good suggestion, I think I will spell it out a bit more even.

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean (hunk 6 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:34]: I think in my head these blank lines separated different categories somehow, but now that I look back on it I think that no blank line is better

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+yoneda.obj Y --yoneda.map snd--> yoneda.obj X
+    |                                |
+   fst                               g
+    |                                |
+    v                                v
+    F ------------ f --------------> G
+``` -/
+def Presheaf.representable : MorphismProperty (Cᵒᵖ ⥤ Type v) :=
+  fun F G f ↦ ∀ ⦃X : C⦄ (g : yoneda.obj X ⟶ G), ∃ (Y : C) (snd : Y ⟶ X)
+    (fst : yoneda.obj Y ⟶ F), IsPullback fst (yoneda.map snd) f g
```
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: I wonder if this should be generalized, i.e. let `Cᵒᵖ ⥤ Type v` just be any category, and `yoneda` just be any functor into it. The definition makes sense, and also a reasonable amount of the API below.

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: (And then have a specialized version for `yoneda`.)

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: `hasRelativePullbacks`, perhaps?

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: (I don't intend this comment to block merging, if it doesn't sounds like a good idea!)

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: I think this sound interesting / fun, so I will try it

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: Actually, maybe it's not so bad.

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean]: Done!

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+In this file, we define a notion of relative representability which works with respect to any
+functor, and not just `yoneda`. The fact that a morphism `f : F ⟶ G` between presheaves is
+representable in the classical case will then be given by `yoneda.relativelyRepresentable f`.
+
+
+
+## Main definitions
+Throughout this file, `F : C ⥤ D` is a functor between categories `C` and `D`.
+
+* We define `relativelyRepresentable` as a `MorphsimProperty`. A morphism `f : X ⟶ Y` in `D` is
```
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:35]:
```suggestion
* We define `relativelyRepresentable` as a `MorphismProperty`. A morphism `f : X ⟶ Y` in `D` is
```

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  said to be relatively representable if for any `g : F.obj a ⟶ Y`, there exists a pullback square
```
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:36]:
```suggestion
  said to be relatively representable with respect to `F` if for any `g : F.obj a ⟶ Y`, there exists a pullback square
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**